share/CMakeJS.cmake

```
@@ -0,0 +1,642 @@
+if(NOT DEFINED CMAKEJS_BINARY_DIR)
+    set(CMAKEJS_BINARY_DIR "${CMAKE_BINARY_DIR}")
+endif()
+
+message (STATUS "\n-- CMakeJS.cmake v${_CMAKEJS_VERSION}")
+
+#[=============================================================================[
+Internal helper (borrowed from CMakeRC).
+]=============================================================================]#
+function(_cmakejs_normalize_path var)
+    set(path "${${var}}")
+    file(TO_CMAKE_PATH "${path}" path)
+    while(path MATCHES "//")
+        string(REPLACE "//" "/" path "${path}")
+    endwhile()
+    string(REGEX REPLACE "/+$" "" path "${path}")
+    set("${var}" "${path}" PARENT_SCOPE)
+endfunction()
+
+#[=======================================================================[
+FindCMakeJs.cmake
+--------
+
+Find the native CMakeJs includes, source, and library
+
+(This codeblock typically belongs in a file named 'FindCMakeJS.cmake' for
+distribution...)
+
+This module defines
+
+::
+
+  CMAKE_JS_INC, where to find node.h, etc.
+  CMAKE_JS_LIB, the libraries required to use CMakeJs.
+  CMAKE_JS_SRC, where to find required *.cpp files, if any,
+  CMAKE_JS_EXECUTABLE, the cmake-js binary (global)
+  CMAKE_JS_NPM_PACKAGE, the cmake-js binary (local)
+
+]=======================================================================]#
+
+# CMAKE_JS_VERSION is defined on all platforms when calling from cmake-js.
+# By checking whether this var is pre-defined, we can determine if we are
+# running from an npm script (via cmake-js), or from CMake directly...
+
+if (NOT DEFINED CMAKE_JS_VERSION) # TODO: Unfortunately, this var is a little too close to 'CMAKE_VERSION' for comfort... Kitware may need it. Should be moved into 'CMAKEJS_VERSION'!
+
+    # ...and if we're calling from CMake directly, we need to set up some vars
+    # that our build step depends on (these are predefined when calling via npm/cmake-js).
+    if(VERBOSE)
+        message(STATUS "CMake Calling...")
+    endif()
+
+    # Check for cmake-js installations
+    find_program(CMAKE_JS_EXECUTABLE
+      NAMES "cmake-js" "cmake-js.exe"
+      PATHS "$ENV{PATH}" "$ENV{ProgramFiles}/cmake-js"
+      DOC "cmake-js system executable binary"
+      REQUIRED
+    )
+    if (NOT CMAKE_JS_EXECUTABLE)
+        message(FATAL_ERROR "cmake-js system installation not found! Please run 'npm -g install cmake-js@latest' and try again.")
+        return()
+    endif()
+
+    _cmakejs_normalize_path(CMAKE_JS_EXECUTABLE)
+    string(REGEX REPLACE "[\r\n\"]" "" CMAKE_JS_EXECUTABLE "${CMAKE_JS_EXECUTABLE}")
+
+    find_program(CMAKE_JS_NPM_PACKAGE
+      NAMES "cmake-js" "cmake-js.exe"
+      PATHS "${CMAKE_CURRENT_SOURCE_DIR}/node_modules/cmake-js/bin"
+      DOC "cmake-js project-local npm package binary"
+      REQUIRED
+    )
+    if (NOT CMAKE_JS_NPM_PACKAGE)
+        message(FATAL_ERROR "cmake-js project-local npm package not found! Please run 'npm install' and try again.")
+        return()
+    endif()
```

```suggestion
    # Check for cmake-js installations
    find_program(CMAKE_JS_EXECUTABLE
      NAMES "cmake-js" "cmake-js.exe"
      PATHS "${CMAKE_CURRENT_LIST_DIR}/../bin"
      DOC "cmake-js project-local npm package binary"
      REQUIRED
    )
    if (NOT CMAKE_JS_EXECUTABLE)
        message(FATAL_ERROR "cmake-js not found! Please run 'npm install' and try again.")
        return()
    endif()
```

I still think that this is sufficient to be able to find a cmake-js binary. If this cmake file has been found from some globally installed cmake-js then it will use the binary from there, or if it is being pulled from the local node_mdoules then it will find it there. The one scenario I don't think this handles is if the user copies this file into their project to modify it, perhaps that can be solved by putting `${CMAKE_CURRENT_LIST_DIR}/..` into a variable which the user can override?